### ISSUE TITLE
Phalcon Model::save() does not respect parameter data types with dynamic updates

### ISSUE THREAD

**Issue by virgofx** [phalcon/cphalcon#12129]
When comparing model records against values from the snapshot during dynamic updates, the constants got juggled such that comparisons were made against the database BIND type instead of the parameter data type.

The affected lines of code are here:
https://github.com/phalcon/cphalcon/blob/master/phalcon/mvc/model.zep#L2410-L2440
```zep
switch bindType {

    case Column::TYPE_BOOLEAN:
        let changed = (boolean) snapshotValue !== (boolean) value;
        break;

    case Column::TYPE_INTEGER:
        let changed = (int) snapshotValue !== (int) value;
        break;

    case Column::TYPE_DECIMAL:
    case Column::TYPE_FLOAT:
        let changed = floatval(snapshotValue) !== floatval(value);
        break;

    case Column::TYPE_DATE:
    case Column::TYPE_VARCHAR:
    case Column::TYPE_DATETIME:
    case Column::TYPE_CHAR:
    case Column::TYPE_TEXT:
    case Column::TYPE_VARCHAR:
    case Column::TYPE_BIGINTEGER:
        let changed = (string) snapshotValue !== (string) value;
        break;

    /**
     * Any other type is not really supported...
     */
    default:
        let changed = value != snapshotValue;
}
```

The `bindType` should be `dataType` where dataType is pulled from the models meta manager.

Additionally, the comparison in `_doLowUpdate()` for changed fields should be copied into the method `getChangedFields()` as this is comparing new vs. old with simple equal checking and not proper equal and identical checking as done in `_doLowUpdate()`

**Comment by virgofx** [comment]
I've confirmed with @andresgutierrez  that changing bindType to the models dataType from the metadata correctly fixes the issue.

**Comment by virgofx** [comment]
Thanks @andresgutierrez - Do we want to update `getChangedFields()` as well for consistency?

**Comment by sergeyklay** [comment]
@virgofx Could you please check the `3.0.x` branch?

**Comment by sergeyklay** [comment]
@virgofx

**Comment by virgofx** [comment]
Rebuilding against 3.0.x now to confirm.

**Comment by virgofx** [comment]
Appears to not work still. Diving in to see exactly differences between codeblock used to verify fix with andres and this one. Will update later today.

**Comment by virgofx** [comment]
Works when building with latest Zephir. Most likely as result of the corresponding type fixes with Andres change in phalcon/zephir - #5809a0f543442cdf29c8e5561b2f26ff8c0f9fe7.

@sergeyklay Can you update Zephir to latest in 3.0.x branch and I'll re-test?

**Comment by andresgutierrez** [comment]
@virgofx We don't have to update Zephir in Phalcon, it always uses the latest one

I'll uploaded the code generated by Zephir

**Comment by virgofx** [comment]
Confirm fixed with:
* Installing 3.0.x via `/ext`
* Installing with latest Zephir build

The `/build/install` doesn't have updated generated files so not fixed when installing Phalcon from `/build`

**Comment by Jurigag** [comment]
Just wait for release.... There is reason why 3.0.x is 3.0.x and fixes aren't directly pushed to master.

When 3.0.1 will be released it will work from `/build`

**Comment by sergeyklay** [comment]
Fixed in the `3.0.x` branch.
